### PR TITLE
ref(models): use DRF APIException as a base for all exceptions

### PR DESCRIPTION
**Pull request by helgi**
## Summary of Changes

Move to using `DeisException` that extends `APIException` (with 400 as a status code) instead of `EnvironmentError`, have `AlreadyExists` extend `APIException` with 409 and `KubeException` and Logger related exceptions get raised up to the view as `ServiceUnavailable`

Additionally using `from e` when re-raising exception to keep contextual information if that is ever required (https://www.python.org/dev/peps/pep-3134/)

https://github.com/tomchristie/django-rest-framework/blob/master/docs/api-guide/exceptions.md has further info on DRF exceptions

## Issue(s) that this PR Closes

ref #612

### CODE REVIEW

File: rootfs/api/models/build.py [hunk 1]
```diff
@@ -43,10 +43,11 @@ def create(self, user, *args, **kwargs):
         try:
             self.app.deploy(new_release)
             return new_release
-        except Exception:
+        except Exception as e:
             if 'new_release' in locals():
                 new_release.delete()
-            raise
+
+            raise DeisException(str(e)) from e
```
Comment (bacongobbler) [inline rootfs/api/models/build.py]: would this mask any important errors? I feel like we'd be losing the exception class that was raised in a traceback if we're converting it straight to a DeisException.

File: rootfs/api/models/certificate.py [hunk 2]
```diff
@@ -179,15 +180,15 @@ def attach_in_kubernetes(self, domain):
...
             self._scheduler._create_secret(app, name, data)
         else:
             # update cert secret to the TLS Ingress format if required
             if secret != data:
                 try:
                     self._scheduler._update_secret(app, name, data)
-                except KubeHTTPException:
-                    raise
+                except KubeException as e:
+                    raise ServiceUnavailable(str(e)) from e
```
Comment (bacongobbler) [inline rootfs/api/models/certificate.py]: I feel like this is out of scope of the scheduler and should be handled in the model instead.

File: rootfs/api/models/certificate.py [hunk 3]
```diff
@@ -220,9 +221,8 @@ def detach(self, *args, **kwargs):
                 # We raise an exception when a secret doesn't exist
                 self._scheduler._get_secret(app, name)
                 self._scheduler._delete_secret(app, name)
-            except KubeHTTPException as e:
-                logger.critical(e)
-                raise EnvironmentError("Could not delete certificate secret {} for application {}".format(name, app))  # noqa
+            except KubeException as e:
+                raise ServiceUnavailable("Could not delete certificate secret {} for application {}".format(name, app)) from e  # noqa
```
Comment (bacongobbler) [inline rootfs/api/models/certificate.py]: same here. The model or template should determine if the error is significant enough to return a SERVICE UNAVAILABLE message, not the scheduler.

File: rootfs/api/models/certificate.py [hunk 4]
```diff
@@ -220,9 +221,8 @@ def detach(self, *args, **kwargs):
                 # We raise an exception when a secret doesn't exist
                 self._scheduler._get_secret(app, name)
                 self._scheduler._delete_secret(app, name)
-            except KubeHTTPException as e:
-                logger.critical(e)
```
Comment (bacongobbler) [inline rootfs/api/models/certificate.py]: any reason why we're removing this message?

File: rootfs/api/models/build.py (hunk 1 above)
Comment (helgi) [inline rootfs/api/models/build.py]: We would not, see the from e portion - also you'd get the error message and proper status code up in the view, which is all it would end up getting piped up into and displayed to the end user as

File: rootfs/api/models/release.py [hunk 5]
```diff
@@ -136,36 +134,38 @@ def previous(self):
...
-            build=prev.build,
-            config=prev.config,
-            summary=summary,
-            source_version='v{}'.format(version)
-        )
-
+    def rollback(self, user, version=None):
         try:
+            # if no version is provided then grab version from object
+            version = (self.version - 1) if version is None else int(version)
```
Comment (bacongobbler) [inline rootfs/api/models/release.py]: can we please retain the old logic in this PR? Let's pull this into another PR for review's sake.

File: rootfs/api/models/certificate.py (hunk 2 above)
Comment (helgi) [inline rootfs/api/models/certificate.py]: But this is in the certificate model that I am re-raising as a different type, not the scheduler. Can you elaborate what you mean?

File: rootfs/api/models/certificate.py (hunk 4 above)
Comment (helgi) [inline rootfs/api/models/certificate.py]: Not consistent with any other exception handling we do

File: rootfs/api/models/release.py [hunk 6]
```diff
@@ -136,36 +134,38 @@ def previous(self):
...
+
             if self.build is not None:
                 self.app.deploy(new_release)
             return new_release
-        except Exception:
+        except Exception as e:
             if 'new_release' in locals():
                 new_release.delete()
-            raise
+            raise DeisException(str(e))
```
Comment (bacongobbler) [inline rootfs/api/models/release.py]: same here as before. Masking the original Exception with a DeisException may mask the original issue (for example a KeyuboardInterrupt or a MemoryError.

File: rootfs/api/models/certificate.py (hunk 3 above)
Comment (helgi) [inline rootfs/api/models/certificate.py]: It doesn't when service unavailable, it has the error message. I think you are just looking at the default message

File: rootfs/api/tests/test_auth.py [hunk 7]
```diff
@@ -248,7 +248,7 @@ def test_passwd(self):
         }
         response = self.client.post(url, submit,
                                     HTTP_AUTHORIZATION='token {}'.format(token))
-        self.assertEqual(response.status_code, 400)
+        self.assertEqual(response.status_code, 401)
```
Comment (bacongobbler) [inline rootfs/api/tests/test_auth.py]: I'd try to keep the responses the same as they were before for compatibility's sake. Is there a specific reason why we want to change this from a 400 to a 401?

File: rootfs/api/models/release.py (hunk 5 above)
Comment (helgi) [inline rootfs/api/models/release.py]: i do not see the point of that, as part of this PR I simplified the view file and this is one of those

File: rootfs/api/tests/test_auth.py (hunk 7 above)
Comment (bacongobbler) [inline rootfs/api/tests/test_auth.py]: If we are going forward with this change can you please make a doc PR to the API reference docs so we can keep it up to date? Thanks!

File: rootfs/api/models/release.py (hunk 6 above)
Comment (helgi) [inline rootfs/api/models/release.py]: It doesn't mask it in any meaningful way, also we never tried catching any of the exceptions you mentions and rarely tried to catch the top level Exception.

We need to not let random exceptions leak out to the user with a massive trace back. We need to handle them

File: rootfs/api/models/build.py (hunk 1 above)
Comment (bacongobbler) [inline rootfs/api/models/build.py]: You'd get the error message and the status code, but not the parent Exception class. We should handle those on a case-by-case basis moving forward rather than mask them all off with a generic Exception class, IMO.

File: rootfs/api/tests/test_auth.py (hunk 7 above)
Comment (helgi) [inline rootfs/api/tests/test_auth.py]: Because it is a failed authentication and not a bad request - we were using the wrong status code

File: rootfs/api/models/certificate.py (hunk 2 above)
Comment (bacongobbler) [inline rootfs/api/models/certificate.py]: Sorry, I meant handled in the views, not in the model. Need more :coffee:

File: rootfs/api/models/build.py (hunk 1 above)
Comment (helgi) [inline rootfs/api/models/build.py]: I think that's the wrong way to do it and lends for a bad user experience. If you are very concerned by this and the nicer UX then we can look at a custom exception handler that can log more ifnromation about parent exception, since I am attaching the originating exception to the re-raised one

File: rootfs/api/models/certificate.py (hunk 2 above)
Comment (helgi) [inline rootfs/api/models/certificate.py]: Views shouldn't be handling exceptions, the way we built it up so far isn't a very clean way

File: rootfs/api/models/build.py (hunk 1 above)
Comment (bacongobbler) [inline rootfs/api/models/build.py]: Yeah, I guess it's two different ways of approaching the same problem. I'm fine with this.

File: rootfs/api/models/certificate.py (hunk 2 above)
Comment (bacongobbler) [inline rootfs/api/models/certificate.py]: I've always been taught that in an MVC world, the views should handle errors and return the appropriate response code, whereas models should retrieve the data and raise errors if it could not retrieve said data. Is this wrong to assume?

Comment (helgi) [inline rootfs/api/models/certificate.py]: Yes and no - There is more than one way to skin that cat.

One thing to pay close attention to here is that we are still raising exceptions up to the View, the difference here is we are curating what exceptions are raised, and they contain all the *required* information for the DRF exception handler to translate that into a `HttpResponse` object. In that sense we have not broken the paradigm you described.

When we use the models across one another (app calling build, build calling app, etc) then they are still Python Exceptions that we can catch and process on, however the Models are not Libraries. The scheduler is a library the Models use, and the Models are the Contract between the View and the 3rd party services (scheduler + k8s, logger, docker)

That's how I have been approach it in any case and how I ended up with what I have now

Comment (bacongobbler) [inline rootfs/api/models/certificate.py]: works for me

File: rootfs/api/models/build.py (hunk 1 above)
Comment (helgi) [inline rootfs/api/models/build.py]: `DeisException` is our generic, like `EnvironmentError` was before - We can always introduce more Exception classes if we want to expose different things up to the API View or generally keep track of things differently

One thing that comes to mind for expansion: `KubeServiceUnavailable` that extends `ServiceUnavailable` and same way `LoggerServiceUnavailable` - Ultimately boils down to a `503` in the View.

File: rootfs/api/models/release.py (hunk 5 above)
Comment (bacongobbler) [inline rootfs/api/models/release.py]: it's just more code creep that I'd rather keep separate, for example changing the function signature and how we retrieve the version number from that is not important to this PR. It just slows down code review because I have to do a context switch in my mind to review this. It looks fine to me though.

Comment (helgi) [inline rootfs/api/models/release.py]: I can look at splitting it out after lunch if it helps

File: rootfs/api/tests/test_auth.py (hunk 7 above)
Comment (bacongobbler) [inline rootfs/api/tests/test_auth.py]: if this was any time other than beta, I'd say this has to be changed, wrong status code or not. Because we're pre-prod though this is fine.

File: rootfs/api/models/release.py (hunk 5 above)
Comment (bacongobbler) [inline rootfs/api/models/release.py]: Nah I reviewed and it's ok. Let's just try to refrain from refactoring parallel code in the future so it's easier on the reviewers :)

File: rootfs/api/tests/test_auth.py (hunk 7 above)
Comment (helgi) [inline rootfs/api/tests/test_auth.py]: I agree there, it was the only reason why I touched that

File: rootfs/api/models/app.py [hunk 8]
```diff
@@ -43,7 +44,7 @@ def validate_app_structure(value):
         if any(int(v) < 0 for v in value.values()):
             raise ValueError("Must be greater than or equal to zero")
     except ValueError as err:
-        raise ValidationError(err)
+        raise ValidationError(str(err))
```
Comment (arschles) [inline rootfs/api/models/app.py]: is this a built-in DRF exception?

Comment (helgi) [inline rootfs/api/models/app.py]: Yeah, ValidationError is both Django and DRF, the latter augments it a bit to better play with its View logic

Comment (bacongobbler) [inline rootfs/api/models/app.py]: https://github.com/tomchristie/django-rest-framework/blob/bb56ca46ed6c07db0146dbdc61c672ff25f127de/rest_framework/exceptions.py#L61-L79
